### PR TITLE
Fix formatting performance problems with .formatter.exs in subdirectories

### PR DESCRIPTION
**Pull request by jonleighton**
When dealing with `.formatter.exs` files that are in subdirectories of the
main project, the behaviour was incorrect: 
https://github.com/elixir-lsp/elixir-ls/issues/361

And also caused performance problems: 
https://github.com/elixir-lsp/elixir-ls/issues/381

A good example is a project I currently work on. We have a top-level
`.formatter.exs`, as well as a `test/.formatter.exs`.

The one in `test/` has the `:inputs` option set to `["**/*.{ex,exs}"]`. 
Before this commit, should_format?/3 would take this pattern and 
concatenate it with the project directory. This would result in 
Path.wildcard/2 globbing for all source files in the entire project, rather
than just under test/. This causes formatting to take about 5 seconds on
our project.

This commit fixes the problem by taking the file to be formatted, and 
finding the `.formatter.exs` file that exists in the closest parent 
directory (so long as that directory is still within the project root). We
then perform the Path.wildcard/2 relative to that directory.

I’ve removed the hack that was in place to support umbrella projects, 
because this is a more general solution to the problem.

On our project, I am now able to format a file in 100ms or so, which isn’t
a bad speed-up!

### CODE REVIEW

File: apps/language_server/lib/language_server/providers/formatting.ex [hunk 1]
```diff
@@ -53,21 +53,49 @@ defmodule ElixirLS.LanguageServer.Providers.Formatting do
...
+    if File.exists?(file_path) do
+      formatter_dir = find_formatter_dir(project_dir, file_path)
+
+      inputs
+      |> Stream.flat_map(&Path.wildcard(Path.join([formatter_dir, &1]), match_dot: true))
+      |> Enum.any?(&(file_path == &1))
+    else
+      # If the file doesn't actually exist on disk, we can't currently check
+      # whether we should format it via Path.wildcard/2, as that operation is
+      # dependent on the filesystem. So just assume that we should.
```
Comment (jonleighton) [inline apps/language_server/lib/language_server/providers/formatting.ex:68]: I am not sure if this is necessary in reality?

I needed it to get the formatting test in `server_test.exs` to pass, because that test doesn't create the file on disk that it is formatting.

Previously it passed because `:inputs` was `nil`, so we didn't use this function def. But I've changed that in this commit so as not to have overlapping `.formatter.exs` files (which is documented as undefined behaviour).

An alternative fix would be to do a `File.write!/2` in the test so that we're dealing with a file that actually exists, but I thought that the file-doesn't-exist case might be a real-world use case? But I don't really know.

File: apps/language_server/lib/language_server/providers/formatting.ex [hunk 2]
```diff
@@ -53,21 +53,49 @@ defmodule ElixirLS.LanguageServer.Providers.Formatting do
...
-    |> Stream.flat_map(&Path.wildcard(&1, match_dot: true))
-    |> Enum.any?(&(file_path == &1))
+    if File.exists?(file_path) do
+      formatter_dir = find_formatter_dir(project_dir, file_path)
+
+      inputs
+      |> Stream.flat_map(&Path.wildcard(Path.join([formatter_dir, &1]), match_dot: true))
+      |> Enum.any?(&(file_path == &1))
+    else
+      # If the file doesn't actually exist on disk, we can't currently check
```
Comment (lukaszsamson) [inline apps/language_server/lib/language_server/providers/formatting.ex:66]: I think this comment is wrong. un unsaved file would have `unsaved:` or `unknown:` schema instead of `file:`. Calling `file_uri |> SourceFile.abs_path_from_uri()` would crash

File: apps/language_server/lib/language_server/providers/formatting.ex [hunk 3]
```diff
@@ -53,21 +53,49 @@ defmodule ElixirLS.LanguageServer.Providers.Formatting do
...
+  # Finds the deepest directory that contains file_path, that also contains a
+  # .formatter.exs. It's possible, though unlikely, that the .formatter.exs we
+  # find is not actually linked to the project_dir via the :subdirectories
+  # option in the top-level .formatter.exs. Currently, that edge case is
+  # glossed over.
+  defp find_formatter_dir(project_dir, file_path) do
+    parts = Path.split(file_path)
+    project_parts = Path.split(project_dir)
+
+    unless List.starts_with?(parts, project_parts) do
```
Comment (lukaszsamson) [inline apps/language_server/lib/language_server/providers/formatting.ex:84]: This may not work well with symlinks

File: apps/language_server/lib/language_server/providers/formatting.ex (hunk 2 above)
Comment (jonleighton) [reply apps/language_server/lib/language_server/providers/formatting.ex:66]: Sounds good -- that was my comment above, I wasn't sure if we needed this but had added to make a test pass. I'll create the file explicitly in the test instead.

File: apps/language_server/lib/language_server/providers/formatting.ex (hunk 3 above)
Comment (jonleighton) [reply apps/language_server/lib/language_server/providers/formatting.ex:84]: I'll have a think about this, although if there's not an obvious solution I'd say that merging this is still far better than the status quo.

**Comment by jonleighton** [comment]
Thanks for the review @lukaszsamson :bow:

**Comment by jonleighton** [comment]
I've updated the branch to address your comments @lukaszsamson

**Comment by jonleighton** [comment]
Ok, I've added one more commit which avoids the filesystem traversal entirely:

* I wrote a library called [`path_glob`](https://github.com/jonleighton/path_glob) that implements matching for globs without interacting with the filesystem
* I've implemented it by writing a [parser](https://github.com/jonleighton/path_glob/blob/main/lib/path_glob/parser.ex) with [`nimble_parsec`](https://github.com/dashbitco/nimble_parsec), so I believe it to be fairly robust. I wrote [a lot of test cases](https://github.com/jonleighton/path_glob/blob/main/test/path_glob_test.exs) and fixed lots of edge cases.
* That said, it's a new library so more testing/review is welcome. But I'm pretty sure it will handle the most common use cases without trouble.
* The new commit just drops in this library.

If there is a concern about maintenance, I'd be happy for the library to be hosted under the `elixir-lsp` org.

**Comment by jonleighton** [comment]
@axelson Thanks for that, I've merged your PR and released v0.1.1 with it.

I'll just note for completeness that I had a chat with José Valim about adding a `Path.match?()` function to Elixir's standard library. He was keen on the idea but said that because `Path.wildcard()` is built on Erlang's `filelib`, it would need to be added to Erlang first. It turns out that there is a private function in `filelib` that provides a parse tree for a glob expression so it wouldn't require implementing a parser, just converting the parse tree into a regex. (I didn't realise this when writing `path_glob`... but it was a fun exploration of `nimble_parsec` anyway.)

I put that in the "maybe one day" basket. I don't think I'm going to make time to do that in the near future but it would be a good addition to the standard library I think and I understand why he doesn't want to just include/maintain a whole new parser in Elixir.

**Comment by axelson** [comment]
@jonleighton Thanks for that write-up and additional detail. I just opened https://github.com/elixir-lsp/elixir-ls/pull/628 once we've tested and merged that I think this PR should be good to go!

**Review by axelson — APPROVED** [review]
This is great! Thanks! :heart:
